// react-static/lib/static/exportRoutes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _os = _interopRequireDefault(require("os"));

var _child_process = require("child_process");

var _chalk = _interopRequireDefault(require("chalk"));

var _utils = require("../utils");

var _fetchSiteData = _interopRequireDefault(require("./fetchSiteData"));

var _fetchRoutes = _interopRequireDefault(require("./fetchRoutes"));

var _plugins = _interopRequireDefault(require("./plugins"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var cores = Math.max(_os["default"].cpus().length, 1);

var _default =
/*#__PURE__*/
function () {
  var _exportRoutes = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee(state) {
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return (0, _fetchSiteData["default"])(state);

          case 2:
            state = _context.sent;
            _context.next = 5;
            return (0, _fetchRoutes["default"])(state);

          case 5:
            state = _context.sent;
            _context.next = 8;
            return buildHTML(state);

          case 8:
            state = _context.sent;
            _context.next = 11;
            return _plugins["default"].afterExport(state);

          case 11:
            state = _context.sent;
            return _context.abrupt("return", state);

          case 13:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  function exportRoutes(_x) {
    return _exportRoutes.apply(this, arguments);
  }

  return exportRoutes;
}();

exports["default"] = _default;

function buildHTML(_x2) {
  return _buildHTML.apply(this, arguments);
}

function _buildHTML() {
  _buildHTML = (0, _asyncToGenerator2["default"])(
  /*#__PURE__*/
  _regenerator["default"].mark(function _callee2(state) {
    var routes, _state$config, paths, maxThreads, threads, htmlProgress, exporters, i, exporterRoutes;

    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            routes = state.routes, _state$config = state.config, paths = _state$config.paths, maxThreads = _state$config.maxThreads;
            (0, _utils.time)(_chalk["default"].green("[\u2713] HTML Exported")); // in case of an absolute path for DIST we must tell node to load the modules
            // from our project root

            if (!paths.DIST.startsWith(paths.ROOT)) {
              process.env.NODE_PATH = paths.NODE_MODULES;

              require('module').Module._initPaths();
            } // Single threaded export


            if (!(maxThreads <= 1)) {
              _context2.next = 9;
              break;
            }

            console.log('Exporting HTML...');
            _context2.next = 7;
            return require('./exportRoutes.sync')["default"](state);

          case 7:
            _context2.next = 18;
            break;

          case 9:
            // Multi-threaded export
            threads = Math.min(cores, maxThreads);
            htmlProgress = (0, _utils.progress)(routes.length);
            console.log("Exporting HTML across ".concat(threads, " threads..."));
            exporters = [];

            for (i = 0; i < threads; i++) {
              exporters.push((0, _child_process.fork)(require.resolve('./exportRoutes.threaded'), [], {
                env: _objectSpread({}, process.env, {
                  REACT_STATIC_THREAD: 'true'
                }),
                stdio: 'inherit'
              }));
            }

            exporterRoutes = exporters.map(function () {
              return [];
            });
            routes.forEach(function (route, i) {
              exporterRoutes[i % exporterRoutes.length].push(route);
            });
            _context2.next = 18;
            return Promise.all(exporters.map(function (exporter, i) {
              var routes = exporterRoutes[i];
              return new Promise(function (resolve, reject) {
                exporter.send(_objectSpread({}, state, {
                  routes: routes
                }));
                exporter.on('message', function (_ref) {
                  var type = _ref.type,
                      payload = _ref.payload;

                  if (type === 'error') {
                    reject(payload);
                  }

                  if (type === 'log') {
                    var _console;

                    (_console = console).log.apply(_console, (0, _toConsumableArray2["default"])(payload));
                  }

                  if (type === 'tick') {
                    htmlProgress.tick();
                  }

                  if (type === 'done') {
                    resolve();
                  }
                });
              });
            }));

          case 18:
            (0, _utils.timeEnd)(_chalk["default"].green("[\u2713] HTML Exported"));
            return _context2.abrupt("return", state);

          case 20:
          case "end":
            return _context2.stop();
        }
      }
    }, _callee2);
  }));
  return _buildHTML.apply(this, arguments);
}
